{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ad1a96df-6110-4762-b66a-de7785796da9",
   "metadata": {},
   "source": [
    "# Assay Binding analysis\n",
    "Analyze how well my PCR primers and probes match the sequences I've found in my samples."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffbdd3ac-2bd4-406b-a7eb-2fad51d7a4b9",
   "metadata": {},
   "source": [
    "## Initialization, configuration and utility functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d845b8fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 1\n",
    "%aimport RCUtils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "40d8036f-16ec-4d80-84b8-7cdf3a6c874a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading primers: HRVPrimers.fasta\n",
      "  ENTng-f (degeneracy 2)\n",
      "  ENTng-r\n",
      "  ENTng-p (degeneracy 8)\n",
      "  ENTrc-f1\n",
      "  ENTrc-f2\n",
      "  ENTrc-r\n",
      "  HRVma-f\n",
      "  HRVma-r\n",
      "  HRVma-p\n",
      "  HRVkaV-fo (degeneracy 2)\n",
      "  HRVkaV-fi\n",
      "  HRVkaV-r (degeneracy 768)\n",
      "  HRVka5-f\n",
      "  HRVka5-ro\n",
      "  HRVka5-ri\n",
      "  HRVbo-f (degeneracy 4)\n",
      "  HRVbo-r\n",
      "  HRVbo-p\n",
      "Read 18 primers\n"
     ]
    }
   ],
   "source": [
    "from IPython.display import display, HTML\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio import Align\n",
    "from Bio.Align import PairwiseAligner, substitution_matrices\n",
    "from Bio.Data import IUPACData\n",
    "import numpy as np\n",
    "import RCUtils\n",
    "import glob\n",
    "from pathlib import Path\n",
    "from collections import defaultdict\n",
    "import pandas as pd\n",
    "\n",
    "# For now this is just Enterovirus primers, but could expand it to other species\n",
    "# We don't want to expand ambiguities because the HRVkaV primers blow up into a huge number.\n",
    "# Instead we use a custom substitution matrix that gives a match if the sets of bases overlap.\n",
    "primers = RCUtils.readPrimers(\"HRVPrimers.fasta\", display=True, expandAmbiguous=False)\n",
    "\n",
    "# Build a substitution matrix where score = +m if sets intersect else -x\n",
    "def iupac_subst(match=1, mismatch=0):\n",
    "    alphabet=\"ACGTRYSWKMBDHVNXI\" # Use X to prevent any matching (for masking)\n",
    "    m = substitution_matrices.Array(alphabet, dims=2, data=np.full((len(alphabet), len(alphabet)), mismatch, float))\n",
    "    lut = {c:set(IUPACData.ambiguous_dna_values.get(c, c)) for c in alphabet}\n",
    "    for i,a in enumerate(alphabet):\n",
    "        for j,b in enumerate(alphabet):\n",
    "            # Match if the sets of bases overlap, or if either is Inosine (matches anything)\n",
    "            if (lut.get(a,{a}) & lut.get(b,{b})) or a == 'I' or b == 'I':\n",
    "                if a != 'X' and b != 'X':\n",
    "                    m[i,j] = match\n",
    "    return m\n",
    "\n",
    "aligner = RCUtils.getPrimerAligner()\n",
    "aligner.substitution_matrix = iupac_subst()\n",
    "\n",
    "def printAlignment(ref, seq, printAlignment=False):\n",
    "    a = aligner.align(ref.seq, seq.seq)[0]\n",
    "    RCUtils.extendAlignment(a)\n",
    "    display(HTML(\"<b>%s aligns to %s[%d:%d]</b> match %d%%\" % \n",
    "        (seq.name, ref.name, a.coordinates[0][0], a.coordinates[0][-1], 100 * a.score / len(seq))))\n",
    "    if printAlignment:\n",
    "        print(a)\n",
    "\n",
    "def printSeqBinding(path, format=None, primers=primers):\n",
    "    if format is None:\n",
    "        format = path.rsplit(\".\",1)[-1]\n",
    "    record = SeqIO.read(path, format)\n",
    "    display(HTML(\"<b>Sequence: %s</b>  len=%d\" % (path, len(record))))\n",
    "    hits = RCUtils.computePrimerHits(record, primers, allowOverlaps=True, aligner=aligner)\n",
    "    for hit in sorted(hits, key=lambda hit: hit.start):\n",
    "        print (\" %s%s len=%d match=%d%% [%d:%d]\" % (hit.primer.id, \" (rev)\" if hit.rev else \"\", len(hit.primer.seq), 100*hit.mr, hit.start+1, hit.end))\n",
    "        if hit.mr < 1:\n",
    "            a = aligner.align(record.seq, hit.primer.rcSeq if hit.rev else hit.primer.seq)[0]\n",
    "            if a.coordinates[1][0] > 0:\n",
    "                if a.coordinates[0][0] == 0:\n",
    "                    print (\"  Primer falls %d bases off the start of the target\" % a.coordinates[1][0])\n",
    "                else:\n",
    "                    print (\"  Primer mismatch in first %d bases\" % a.coordinates[1][0])\n",
    "            pt = len(hit.primer.seq) - a.coordinates[1][-1]\n",
    "            if pt > 0:\n",
    "                if a.coordinates[0][-1] == len(record):\n",
    "                    print (\"  Primer falls %d bases off the end of the target\" % pt)\n",
    "                else:\n",
    "                    print (\"  Primer mismatch in the last %d bases\" % pt)\n",
    "            RCUtils.extendAlignment(a)\n",
    "            print(a)\n",
    "    return record\n",
    "\n",
    "# Show a table of primer match scores for each sequence\n",
    "# If collapse is True, then include fwd, rev and probe in the same\n",
    "# group. May be confusing when only some of the set are present.\n",
    "def summaryTable(pathGlob, format=\"fastq\", collapse=True, primers=primers):\n",
    "    table = dict()\n",
    "    for path in sorted(glob.glob(pathGlob)):\n",
    "        try:\n",
    "            record = SeqIO.read(path, format)\n",
    "        except ValueError as e:\n",
    "            raise ValueError(f\"Error processing file {path}\") from e\n",
    "\n",
    "        # Note that unlike above, we use the default behavior of removing redundant hits\n",
    "        # here so we take the maximum score for variants like ENTrc-f1 and ENTrc-f2 while\n",
    "        # still multiplying scores for distinct primers like ENTrc-f1 and ENTrc-r.\n",
    "        hits = RCUtils.computePrimerHits(record, primers, allowOverlaps=True, aligner=aligner)\n",
    "        scores = dict()\n",
    "        for hit in hits:\n",
    "            # Combine scores for multiple primers in the set\n",
    "            pe = \"-\" if collapse else \".\"\n",
    "            pname = hit.primer.id.rsplit(pe,1)[0]\n",
    "            if pname in scores:\n",
    "                scores[pname] *= hit.mr\n",
    "            else:\n",
    "                scores[pname] = hit.mr\n",
    "        \n",
    "        # Get the base filename without the suffix\n",
    "        file = Path(path).stem\n",
    "        table[file] = scores\n",
    "\n",
    "    pd.set_option('display.max_rows', 500)\n",
    "    df = pd.DataFrame.from_dict(table, orient='index')\n",
    "    df.fillna(0, inplace=True)\n",
    "    df *= 100\n",
    "    df = df.round(0).astype(int)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f40d2a13",
   "metadata": {},
   "source": [
    "## Summary\n",
    "\n",
    "Note that the myseqs sequences are ampicons generated from these same (or similar) primers and should have the primers trimmed and so will not match the primers they were generated from."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "f93e2355",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HRVbo-f</th>\n",
       "      <th>HRVma-f</th>\n",
       "      <th>HRVma-p</th>\n",
       "      <th>HRVka5-ri</th>\n",
       "      <th>HRVbo-p</th>\n",
       "      <th>HRVkaV-fo</th>\n",
       "      <th>ENTng-f</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S130-RVA-58</th>\n",
       "      <td>100</td>\n",
       "      <td>89</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S142-RVA-62</th>\n",
       "      <td>100</td>\n",
       "      <td>89</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S147-RVC-44</th>\n",
       "      <td>100</td>\n",
       "      <td>89</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S148-RVA-68</th>\n",
       "      <td>100</td>\n",
       "      <td>89</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S153-RVB-27</th>\n",
       "      <td>100</td>\n",
       "      <td>89</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S28-RVA-23</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S44-RVA-56</th>\n",
       "      <td>100</td>\n",
       "      <td>89</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S48-RVC-1</th>\n",
       "      <td>100</td>\n",
       "      <td>89</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S59-RVA-77</th>\n",
       "      <td>100</td>\n",
       "      <td>89</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S65-RVA-54</th>\n",
       "      <td>100</td>\n",
       "      <td>89</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S69-RVC-45</th>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S81-RVA-31</th>\n",
       "      <td>100</td>\n",
       "      <td>89</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S91-RVC-45</th>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S94-RVC-1</th>\n",
       "      <td>100</td>\n",
       "      <td>89</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             HRVbo-f  HRVma-f  HRVma-p  HRVka5-ri  HRVbo-p  HRVkaV-fo  ENTng-f\n",
       "S130-RVA-58      100       89      100        100      100        100      100\n",
       "S142-RVA-62      100       89      100        100      100        100      100\n",
       "S147-RVC-44      100       89      100        100      100        100      100\n",
       "S148-RVA-68      100       89      100        100      100        100      100\n",
       "S153-RVB-27      100       89      100        100      100        100      100\n",
       "S28-RVA-23       100      100      100        100      100        100      100\n",
       "S44-RVA-56       100       89      100        100      100        100      100\n",
       "S48-RVC-1        100       89      100        100      100        100      100\n",
       "S59-RVA-77       100       89      100        100      100        100      100\n",
       "S65-RVA-54       100       89      100        100      100        100      100\n",
       "S69-RVC-45       100       95      100        100      100        100      100\n",
       "S81-RVA-31       100       89      100        100      100        100      100\n",
       "S91-RVC-45       100       95      100        100      100        100      100\n",
       "S94-RVC-1        100       89      100        100      100        100      100"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summaryTable(\"myseqs/*-RV*.fastq\", collapse=False)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4851c1da",
   "metadata": {},
   "source": [
    "Overall we see HRVMa is a pretty poor match for everything except S28, which matches qPCR experimental results. Everything else is generally good."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "d076b627",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENTrc</th>\n",
       "      <th>HRVka5</th>\n",
       "      <th>HRVbo</th>\n",
       "      <th>HRVma</th>\n",
       "      <th>HRVkaV</th>\n",
       "      <th>ENTng</th>\n",
       "      <th>random</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A1</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A10</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A100</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A101</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A102</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>86</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A103</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>86</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A104</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A105</th>\n",
       "      <td>82</td>\n",
       "      <td>73</td>\n",
       "      <td>93</td>\n",
       "      <td>71</td>\n",
       "      <td>95</td>\n",
       "      <td>83</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A106</th>\n",
       "      <td>78</td>\n",
       "      <td>100</td>\n",
       "      <td>41</td>\n",
       "      <td>80</td>\n",
       "      <td>61</td>\n",
       "      <td>66</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A11</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A12</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A13</th>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A15</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A16</th>\n",
       "      <td>95</td>\n",
       "      <td>87</td>\n",
       "      <td>84</td>\n",
       "      <td>90</td>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A18</th>\n",
       "      <td>100</td>\n",
       "      <td>94</td>\n",
       "      <td>75</td>\n",
       "      <td>79</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A19</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A1B</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A2</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A20</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A21</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>86</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A22</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A23</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>81</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A24</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A25</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A28</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A29</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A30</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>86</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A31</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A32</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A33</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>74</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A34</th>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A36</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A38</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A39</th>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A40</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A41</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>89</td>\n",
       "      <td>83</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A43</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A45</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A46</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A47</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A49</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A50</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A51</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>86</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A53</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A54</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A55</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>83</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A56</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A57</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A58</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A59</th>\n",
       "      <td>95</td>\n",
       "      <td>90</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A60</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A61</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A62</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A63</th>\n",
       "      <td>90</td>\n",
       "      <td>86</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A64</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>83</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A65</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>86</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A66</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>83</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A67</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>90</td>\n",
       "      <td>83</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A68</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A7</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A71</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A73</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>89</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A74</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A75</th>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>84</td>\n",
       "      <td>83</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A76</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A77</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>83</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A78</th>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>86</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A8</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A80</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A81</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A82</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>89</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A85</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A88</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A89</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A9</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A90</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A94</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A96</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>84</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B101</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B102</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>83</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B103</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>83</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B104</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B14</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B17</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>100</td>\n",
       "      <td>83</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B26</th>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B27</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B3</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>83</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B35</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>84</td>\n",
       "      <td>83</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B37</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B4</th>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B42</th>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>100</td>\n",
       "      <td>84</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B48</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B5</th>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>100</td>\n",
       "      <td>84</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B52</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B6</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B69</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>80</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B70</th>\n",
       "      <td>100</td>\n",
       "      <td>94</td>\n",
       "      <td>75</td>\n",
       "      <td>89</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B72</th>\n",
       "      <td>100</td>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B79</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>100</td>\n",
       "      <td>83</td>\n",
       "      <td>100</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B83</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>83</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B84</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>83</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B86</th>\n",
       "      <td>90</td>\n",
       "      <td>75</td>\n",
       "      <td>83</td>\n",
       "      <td>70</td>\n",
       "      <td>69</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B91</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>66</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B92</th>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>80</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B93</th>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>100</td>\n",
       "      <td>83</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B97</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>80</td>\n",
       "      <td>100</td>\n",
       "      <td>83</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B99</th>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>100</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C1</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>86</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C10</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>86</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C11</th>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>96</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C15</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C2</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>96</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C22</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C23</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>76</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C3</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C35</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>87</td>\n",
       "      <td>85</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C39</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>64</td>\n",
       "      <td>76</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C4</th>\n",
       "      <td>95</td>\n",
       "      <td>89</td>\n",
       "      <td>94</td>\n",
       "      <td>85</td>\n",
       "      <td>69</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C40</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>96</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C41</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>86</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C42</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>76</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C43</th>\n",
       "      <td>86</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>76</td>\n",
       "      <td>76</td>\n",
       "      <td>94</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C44</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C45</th>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>80</td>\n",
       "      <td>90</td>\n",
       "      <td>96</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C49</th>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>76</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C5</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>90</td>\n",
       "      <td>96</td>\n",
       "      <td>83</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C50</th>\n",
       "      <td>95</td>\n",
       "      <td>90</td>\n",
       "      <td>78</td>\n",
       "      <td>89</td>\n",
       "      <td>87</td>\n",
       "      <td>90</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C51</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>87</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>83</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C54</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>87</td>\n",
       "      <td>95</td>\n",
       "      <td>80</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C57</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>86</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C6</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>86</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C7</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>64</td>\n",
       "      <td>86</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C8</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C9</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>87</td>\n",
       "      <td>80</td>\n",
       "      <td>69</td>\n",
       "      <td>100</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-B100</th>\n",
       "      <td>0</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C34</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>96</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ENTrc  HRVka5  HRVbo  HRVma  HRVkaV  ENTng  random\n",
       "Rhinovirus-A1      100     100    100     95     100    100       0\n",
       "Rhinovirus-A10     100     100    100     80     100    100      80\n",
       "Rhinovirus-A100    100     100    100     85     100    100       0\n",
       "Rhinovirus-A101     95      95    100     80     100    100       0\n",
       "Rhinovirus-A102     95      95    100     86     100    100       0\n",
       "Rhinovirus-A103    100     100    100     86     100    100       0\n",
       "Rhinovirus-A104    100     100    100     85     100    100       0\n",
       "Rhinovirus-A105     82      73     93     71      95     83      86\n",
       "Rhinovirus-A106     78     100     41     80      61     66      61\n",
       "Rhinovirus-A11      95      95    100     85     100    100      80\n",
       "Rhinovirus-A12     100     100    100     80     100    100       0\n",
       "Rhinovirus-A13      90      90    100     95     100    100       0\n",
       "Rhinovirus-A15     100     100    100     85     100    100       0\n",
       "Rhinovirus-A16      95      87     84     90      86     86       0\n",
       "Rhinovirus-A18     100      94     75     79      95     95       0\n",
       "Rhinovirus-A19     100     100    100     95     100    100       0\n",
       "Rhinovirus-A1B     100     100    100    100     100    100       0\n",
       "Rhinovirus-A2      100     100    100     90     100    100      80\n",
       "Rhinovirus-A20      95      95    100     85     100    100      80\n",
       "Rhinovirus-A21     100     100    100     86     100    100       0\n",
       "Rhinovirus-A22     100     100    100     95     100    100       0\n",
       "Rhinovirus-A23     100     100    100     81     100    100       0\n",
       "Rhinovirus-A24     100     100    100     85     100    100      80\n",
       "Rhinovirus-A25      95      95    100     85     100    100      56\n",
       "Rhinovirus-A28      95      95    100     85     100    100       0\n",
       "Rhinovirus-A29     100     100    100     85     100    100      80\n",
       "Rhinovirus-A30     100     100    100     86     100    100       0\n",
       "Rhinovirus-A31     100     100    100     80     100    100       0\n",
       "Rhinovirus-A32     100     100    100     90     100    100       0\n",
       "Rhinovirus-A33     100     100     74     85     100    100      80\n",
       "Rhinovirus-A34      90      90    100     80     100    100      80\n",
       "Rhinovirus-A36     100     100    100     80     100    100       0\n",
       "Rhinovirus-A38     100     100    100     85     100    100       0\n",
       "Rhinovirus-A39      90      90    100     90     100    100       0\n",
       "Rhinovirus-A40     100     100     80     85     100    100      80\n",
       "Rhinovirus-A41      95      95    100     89      83    100       0\n",
       "Rhinovirus-A43      95      95    100     95     100    100       0\n",
       "Rhinovirus-A45     100     100    100     90     100    100      80\n",
       "Rhinovirus-A46      95      95    100     80     100    100       0\n",
       "Rhinovirus-A47     100     100    100     85     100    100       0\n",
       "Rhinovirus-A49     100     100    100     90     100    100      74\n",
       "Rhinovirus-A50      95      95    100     80     100    100       0\n",
       "Rhinovirus-A51      95      95    100     86     100    100       0\n",
       "Rhinovirus-A53     100     100    100     90     100    100       0\n",
       "Rhinovirus-A54     100     100    100     85     100    100       0\n",
       "Rhinovirus-A55     100     100    100     90     100     83       0\n",
       "Rhinovirus-A56      95      95    100     85     100    100       0\n",
       "Rhinovirus-A57     100     100    100     85     100    100       0\n",
       "Rhinovirus-A58     100     100    100     85     100    100       0\n",
       "Rhinovirus-A59      95      90     89     89      90     90      80\n",
       "Rhinovirus-A60     100     100    100     90     100    100       0\n",
       "Rhinovirus-A61      95      95    100     95     100    100       0\n",
       "Rhinovirus-A62     100     100    100     85     100    100      70\n",
       "Rhinovirus-A63      90      86     89     89      90     90       0\n",
       "Rhinovirus-A64     100     100    100     95      83    100       0\n",
       "Rhinovirus-A65      95      95    100     86     100    100       0\n",
       "Rhinovirus-A66     100     100    100     85      83    100       0\n",
       "Rhinovirus-A67      95      95    100     90      83    100       0\n",
       "Rhinovirus-A68      95      95    100     85     100    100       0\n",
       "Rhinovirus-A7      100     100    100     85     100    100       0\n",
       "Rhinovirus-A71     100     100    100     90     100    100       0\n",
       "Rhinovirus-A73     100     100    100     89     100    100       0\n",
       "Rhinovirus-A74     100     100    100     90     100    100      80\n",
       "Rhinovirus-A75      90      90    100     84      83    100       0\n",
       "Rhinovirus-A76     100     100    100     85     100    100      80\n",
       "Rhinovirus-A77     100     100    100     85      83    100       0\n",
       "Rhinovirus-A78      90      90    100     86     100    100       0\n",
       "Rhinovirus-A8      100     100    100     90     100    100       0\n",
       "Rhinovirus-A80     100     100    100     85     100    100       0\n",
       "Rhinovirus-A81      95      95    100     90     100    100       0\n",
       "Rhinovirus-A82     100     100    100     89     100    100      80\n",
       "Rhinovirus-A85     100     100    100     85     100    100      87\n",
       "Rhinovirus-A88     100     100    100     85     100    100       0\n",
       "Rhinovirus-A89     100     100    100     85     100    100       0\n",
       "Rhinovirus-A9      100     100    100     95     100    100       0\n",
       "Rhinovirus-A90     100     100    100     80     100    100      80\n",
       "Rhinovirus-A94     100     100    100     95     100    100       0\n",
       "Rhinovirus-A96      95      95    100     84     100    100       0\n",
       "Rhinovirus-B101    100     100    100     95     100    100       0\n",
       "Rhinovirus-B102     95      95    100     95      83    100       0\n",
       "Rhinovirus-B103    100     100    100     95      83    100       0\n",
       "Rhinovirus-B104    100     100     80     95     100    100       0\n",
       "Rhinovirus-B14      95      95    100     95     100    100       0\n",
       "Rhinovirus-B17     100     100    100     80     100     83       0\n",
       "Rhinovirus-B26      86      86    100     80     100    100       0\n",
       "Rhinovirus-B27     100     100    100    100     100    100       0\n",
       "Rhinovirus-B3      100     100    100     95      83    100       0\n",
       "Rhinovirus-B35     100     100    100     84      83    100      80\n",
       "Rhinovirus-B37     100     100    100     80     100    100       0\n",
       "Rhinovirus-B4       86      86    100     95     100    100      80\n",
       "Rhinovirus-B42      86      86    100     84     100    100       0\n",
       "Rhinovirus-B48     100     100    100     95     100    100      64\n",
       "Rhinovirus-B5       86      86    100     84     100    100      80\n",
       "Rhinovirus-B52      95      95    100     95     100    100      80\n",
       "Rhinovirus-B6      100     100    100     95     100    100      80\n",
       "Rhinovirus-B69     100     100     80     80     100    100       0\n",
       "Rhinovirus-B70     100      94     75     89      95     95       0\n",
       "Rhinovirus-B72     100      94     94     94     100    100       0\n",
       "Rhinovirus-B79     100     100     80    100      83    100      64\n",
       "Rhinovirus-B83     100     100    100    100      83    100       0\n",
       "Rhinovirus-B84      95      95    100    100      83    100      80\n",
       "Rhinovirus-B86      90      75     83     70      69    100      80\n",
       "Rhinovirus-B91     100     100    100     80      66    100       0\n",
       "Rhinovirus-B92      90      90     80    100     100    100      80\n",
       "Rhinovirus-B93      90      90    100     80     100     83       0\n",
       "Rhinovirus-B97      95      95     80    100      83    100       0\n",
       "Rhinovirus-B99      86      86    100     95     100    100      80\n",
       "Rhinovirus-C1      100     100    100     86     100    100       0\n",
       "Rhinovirus-C10     100     100    100     86     100    100      80\n",
       "Rhinovirus-C11      90      90    100     80      96    100      80\n",
       "Rhinovirus-C15     100     100    100     90     100    100      64\n",
       "Rhinovirus-C2       95      95    100     80      96    100       0\n",
       "Rhinovirus-C22     100     100    100     90     100    100       0\n",
       "Rhinovirus-C23     100     100    100     76     100    100      80\n",
       "Rhinovirus-C3      100     100    100     90     100    100       0\n",
       "Rhinovirus-C35     100     100     87     85     100    100      80\n",
       "Rhinovirus-C39     100     100     64     76     100    100       0\n",
       "Rhinovirus-C4       95      89     94     85      69    100      80\n",
       "Rhinovirus-C40     100     100    100     80      96    100       0\n",
       "Rhinovirus-C41      95      95    100     86     100    100       0\n",
       "Rhinovirus-C42     100     100    100     76     100    100       0\n",
       "Rhinovirus-C43      86      89     89     76      76     94     100\n",
       "Rhinovirus-C44      95      95    100     80     100    100       0\n",
       "Rhinovirus-C45      90      90     80     90      96    100       0\n",
       "Rhinovirus-C49      95     100    100     76     100    100     100\n",
       "Rhinovirus-C5       95      95    100     90      96     83       0\n",
       "Rhinovirus-C50      95      90     78     89      87     90       0\n",
       "Rhinovirus-C51      95      95     87     90     100     83      80\n",
       "Rhinovirus-C54     100     100     87     95      80    100       0\n",
       "Rhinovirus-C57     100     100     80     86     100    100      80\n",
       "Rhinovirus-C6      100     100    100     86     100    100       0\n",
       "Rhinovirus-C7       95      95     64     86     100    100       0\n",
       "Rhinovirus-C8      100     100    100     80     100    100       0\n",
       "Rhinovirus-C9      100     100     87     80      69    100      80\n",
       "Rhinovirus-B100      0      83     83     83      83      0       0\n",
       "Rhinovirus-C34       0       0      0      0      96      0      64"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summaryTable(\"refseq/Rhinovirus-*.fasta\", \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33b84307-c19f-49e3-995b-963c052eaf1f",
   "metadata": {},
   "source": [
    "# My sequence details\n",
    "## Rhinovirus A-23: S28"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "81b39a7f-aa40-4d85-a2f4-ace850331734",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b>Sequence: refseq/Rhinovirus-A23.gb</b>  len=7025"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " ENTng-p (rev) len=18 match=100% [2:19]\n",
      " HRVma-r len=20 match=90% [2:21]\n",
      "target            1 CAAAGTAGTTGGTCCCGTCC 21\n",
      "                  0 |||||||||.||||||.||| 20\n",
      "query             0 CAAAGTAGTCGGTCCCATCC 20\n",
      "\n",
      " ENTrc-f1 len=21 match=100% [86:106]\n",
      " HRVka5-f len=21 match=100% [86:106]\n",
      " HRVbo-f len=15 match=100% [276:290]\n",
      " HRVma-f len=19 match=100% [321:339]\n",
      " HRVma-p len=18 match=100% [359:376]\n",
      " HRVka5-ri (rev) len=18 match=100% [359:376]\n",
      " HRVbo-p len=18 match=100% [359:376]\n",
      " HRVkaV-fo len=20 match=100% [364:383]\n",
      " ENTng-f len=19 match=100% [365:383]\n",
      " HRVma-r (rev) len=20 match=90% [445:464]\n",
      "target          444 GGACGGGACCAACTACTTTG 464\n",
      "                  0 |||.||||||.|||||||||  20\n",
      "query             0 GGATGGGACCGACTACTTTG  20\n",
      "\n",
      " ENTng-p len=18 match=100% [447:464]\n",
      " ENTrc-r (rev) len=21 match=100% [456:476]\n",
      " HRVka5-ro (rev) len=23 match=100% [456:478]\n",
      " ENTng-r (rev) len=22 match=100% [457:478]\n",
      " HRVkaV-fi len=22 match=100% [457:478]\n",
      " HRVbo-r (rev) len=19 match=100% [460:478]\n",
      " HRVkaV-r (rev) len=23 match=100% [970:992]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>Sequence: myseqs/S28-RVA-23.fastq</b>  len=348"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " HRVbo-f len=15 match=100% [169:183]\n",
      " HRVma-f len=19 match=100% [214:232]\n",
      " HRVma-p len=18 match=100% [252:269]\n",
      " HRVka5-ri (rev) len=18 match=100% [252:269]\n",
      " HRVbo-p len=18 match=100% [252:269]\n",
      " HRVkaV-fo len=20 match=100% [257:276]\n",
      " ENTng-f len=19 match=100% [258:276]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>S28-RVA-23 aligns to DQ473497[106:455]</b> match 95%"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ref = printSeqBinding(\"refseq/Rhinovirus-A23.gb\", \"gb\")\n",
    "seq = printSeqBinding(\"myseqs/S28-RVA-23.fastq\")\n",
    "printAlignment(ref, seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33b6d070",
   "metadata": {},
   "source": [
    "## Rhinovirus A-31: S81"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "45897164",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b>Sequence: myseqs/S81-RVA-31.fastq</b>  len=350"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " HRVbo-f len=15 match=100% [170:184]\n",
      " HRVma-f len=19 match=89% [216:234]\n",
      "target          215 TTGACAAGGTGTGAAGAGC 234\n",
      "                  0 |.||||.||||||||||||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [254:271]\n",
      " HRVka5-ri (rev) len=18 match=100% [254:271]\n",
      " HRVbo-p len=18 match=100% [254:271]\n",
      " HRVkaV-fo len=20 match=100% [259:278]\n",
      " ENTng-f len=19 match=100% [260:278]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>Sequence: refseq/Rhinovirus-A31.fasta</b>  len=7131"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " ENTrc-f1 len=21 match=100% [162:182]\n",
      " HRVka5-f len=21 match=100% [162:182]\n",
      " HRVbo-f len=15 match=100% [352:366]\n",
      " HRVma-f len=19 match=89% [398:416]\n",
      "target          397 TTGACAAGGTGTGAAGAGC 416\n",
      "                  0 |.||||.||||||||||||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [436:453]\n",
      " HRVka5-ri (rev) len=18 match=100% [436:453]\n",
      " HRVbo-p len=18 match=100% [436:453]\n",
      " HRVkaV-fo len=20 match=100% [441:460]\n",
      " ENTng-f len=19 match=100% [442:460]\n",
      " HRVma-r (rev) len=20 match=90% [522:541]\n",
      "target          521 GGACGGGACCAACTACTTTG 541\n",
      "                  0 |||.||||||.|||||||||  20\n",
      "query             0 GGATGGGACCGACTACTTTG  20\n",
      "\n",
      " ENTng-p len=18 match=100% [524:541]\n",
      " ENTrc-r (rev) len=21 match=100% [533:553]\n",
      " HRVka5-ro (rev) len=23 match=100% [533:555]\n",
      " ENTng-r (rev) len=22 match=100% [534:555]\n",
      " HRVkaV-fi len=22 match=100% [534:555]\n",
      " HRVbo-r (rev) len=19 match=100% [537:555]\n",
      " random (rev) len=15 match=80% [817:831]\n",
      "target          816 AATCTCCTACAGTAG 831\n",
      "                  0 |||.|||..||||||  15\n",
      "query             0 AATGTCCCCCAGTAG  15\n",
      "\n",
      " HRVkaV-r (rev) len=23 match=100% [1047:1069]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>S81-RVA-31 aligns to FJ445126.1[182:532]</b> match 97%"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seq = printSeqBinding(\"myseqs/S81-RVA-31.fastq\")\n",
    "ref = printSeqBinding(\"refseq/Rhinovirus-A31.fasta\", \"fasta\")\n",
    "printAlignment(ref, seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66dae935-f608-470f-9ee7-8f851ff05f71",
   "metadata": {},
   "source": [
    "## Rhinovirus A-56: S44"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "13caa15c-bb46-4185-a0bc-0d619606df49",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b>Sequence: myseqs/S44-RVA-56.fastq</b>  len=350"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " HRVbo-f len=15 match=100% [171:185]\n",
      " HRVma-f len=19 match=89% [216:234]\n",
      "target          215 TTGACAAGGTGTGAAGAGC 234\n",
      "                  0 |.||||.||||||||||||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [254:271]\n",
      " HRVka5-ri (rev) len=18 match=100% [254:271]\n",
      " HRVbo-p len=18 match=100% [254:271]\n",
      " HRVkaV-fo len=20 match=100% [259:278]\n",
      " ENTng-f len=19 match=100% [260:278]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>Sequence: refseq/Rhinovirus-A56.gb</b>  len=7136"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " ENTrc-f1 len=21 match=95% [161:181]\n",
      "target          160 CAAGCACTTCTGTCTCCCCGG 181\n",
      "                  0 |||||||||||||.|||||||  21\n",
      "query             0 CAAGCACTTCTGTTTCCCCGG  21\n",
      "\n",
      " HRVka5-f len=21 match=95% [161:181]\n",
      "target          160 CAAGCACTTCTGTCTCCCCGG 181\n",
      "                  0 |||||||||||||.|||||||  21\n",
      "query             0 CAAGCACTTCTGTTTCCCCGG  21\n",
      "\n",
      " HRVbo-f len=15 match=100% [352:366]\n",
      " HRVma-f len=19 match=89% [397:415]\n",
      "target          396 TTGACAAGGTGTGAARAGC 415\n",
      "                  0 |.||||.||||||||.|||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [435:452]\n",
      " HRVka5-ri (rev) len=18 match=100% [435:452]\n",
      " HRVbo-p len=18 match=100% [435:452]\n",
      " HRVkaV-fo len=20 match=100% [440:459]\n",
      " ENTng-f len=19 match=100% [441:459]\n",
      " HRVma-r (rev) len=20 match=95% [521:540]\n",
      "target          520 GGATGGGACCAACTACTTTG 540\n",
      "                  0 ||||||||||.|||||||||  20\n",
      "query             0 GGATGGGACCGACTACTTTG  20\n",
      "\n",
      " ENTng-p len=18 match=100% [523:540]\n",
      " ENTrc-r (rev) len=21 match=100% [532:552]\n",
      " HRVka5-ro (rev) len=23 match=100% [532:554]\n",
      " ENTng-r (rev) len=22 match=100% [533:554]\n",
      " HRVkaV-fi len=22 match=100% [533:554]\n",
      " HRVbo-r (rev) len=19 match=100% [536:554]\n",
      " HRVkaV-r (rev) len=23 match=100% [1049:1071]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>S44-RVA-56 aligns to FJ445140[181:531]</b> match 96%"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seq = printSeqBinding(\"myseqs/S44-RVA-56.fastq\")\n",
    "ref = printSeqBinding(\"refseq/Rhinovirus-A56.gb\", \"gb\")\n",
    "printAlignment(ref, seq)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9c46e05e",
   "metadata": {},
   "source": [
    "Here we see the critical final base of the HRVMA-f primer mismatches our S48 RVC-1 sequence, and there are three other mismatches. This explains why I couldn't get S48 to test positive via HRV qPCR. However, it does seem to be a perfect match for ENTng, so I still don't know why I had so much trouble with that assay being unreliable for this sample. The ENT rc and HRVka primer sets match perfectly."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb112359",
   "metadata": {},
   "source": [
    "## Rhinovirus A-54: S65"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2a635f5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b>Sequence: myseqs/S65-RVA-54.fastq</b>  len=350"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " HRVbo-f len=15 match=100% [170:184]\n",
      " HRVma-f len=19 match=89% [216:234]\n",
      "target          215 TTGACAAGGTGTGAAGAGC 234\n",
      "                  0 |.||||.||||||||||||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [254:271]\n",
      " HRVka5-ri (rev) len=18 match=100% [254:271]\n",
      " HRVbo-p len=18 match=100% [254:271]\n",
      " HRVkaV-fo len=20 match=100% [259:278]\n",
      " ENTng-f len=19 match=100% [260:278]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>Sequence: refseq/Rhinovirus-A54.fasta</b>  len=7134"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " ENTrc-f1 len=21 match=100% [162:182]\n",
      " HRVka5-f len=21 match=100% [162:182]\n",
      " HRVbo-f len=15 match=100% [352:366]\n",
      " HRVma-f len=19 match=89% [398:416]\n",
      "target          397 TTGACAAGGTGTGAAGAGC 416\n",
      "                  0 |.||||.||||||||||||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [436:453]\n",
      " HRVka5-ri (rev) len=18 match=100% [436:453]\n",
      " HRVbo-p len=18 match=100% [436:453]\n",
      " HRVkaV-fo len=20 match=100% [441:460]\n",
      " ENTng-f len=19 match=100% [442:460]\n",
      " HRVma-r (rev) len=20 match=95% [522:541]\n",
      "target          521 GGATGGGACCAACTACTTTG 541\n",
      "                  0 ||||||||||.|||||||||  20\n",
      "query             0 GGATGGGACCGACTACTTTG  20\n",
      "\n",
      " ENTng-p len=18 match=100% [524:541]\n",
      " ENTrc-r (rev) len=21 match=100% [533:553]\n",
      " HRVka5-ro (rev) len=23 match=100% [533:555]\n",
      " ENTng-r (rev) len=22 match=100% [534:555]\n",
      " HRVkaV-fi len=22 match=100% [534:555]\n",
      " HRVbo-r (rev) len=19 match=100% [537:555]\n",
      " HRVkaV-r (rev) len=23 match=100% [1047:1069]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>S65-RVA-54 aligns to FJ445138.1[182:532]</b> match 95%"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seq = printSeqBinding(\"myseqs/S65-RVA-54.fastq\")\n",
    "ref = printSeqBinding(\"refseq/Rhinovirus-A54.fasta\", \"fasta\")\n",
    "printAlignment(ref, seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f7ef80d",
   "metadata": {},
   "source": [
    "## Rhinovirus A-58: S130"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cfc3b2dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b>Sequence: refseq/Rhinovirus-A58.fasta</b>  len=7140"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " ENTrc-f1 len=21 match=100% [164:184]\n",
      " HRVka5-f len=21 match=100% [164:184]\n",
      " HRVbo-f len=15 match=100% [356:370]\n",
      " HRVma-f len=19 match=89% [402:420]\n",
      "target          401 TTGACATGGTGTGAAGAGC 420\n",
      "                  0 |.||||.||||||||||||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [440:457]\n",
      " HRVka5-ri (rev) len=18 match=100% [440:457]\n",
      " HRVbo-p len=18 match=100% [440:457]\n",
      " HRVkaV-fo len=20 match=100% [445:464]\n",
      " ENTng-f len=19 match=100% [446:464]\n",
      " HRVma-r (rev) len=20 match=95% [526:545]\n",
      "target          525 GGATGGGACCAACTACTTTG 545\n",
      "                  0 ||||||||||.|||||||||  20\n",
      "query             0 GGATGGGACCGACTACTTTG  20\n",
      "\n",
      " ENTng-p len=18 match=100% [528:545]\n",
      " ENTrc-r (rev) len=21 match=100% [537:557]\n",
      " HRVka5-ro (rev) len=23 match=100% [537:559]\n",
      " ENTng-r (rev) len=22 match=100% [538:559]\n",
      " HRVkaV-fi len=22 match=100% [538:559]\n",
      " HRVbo-r (rev) len=19 match=100% [541:559]\n",
      " HRVkaV-r (rev) len=23 match=100% [1054:1076]\n",
      " random (rev) len=15 match=80% [2774:2788]\n",
      "  Primer mismatch in first 1 bases\n",
      "target         2772 TATGTACCCCCAGGAG 2788\n",
      "                  0 .||||-|||||||.||   16\n",
      "query             0 AATGT-CCCCCAGTAG   15\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>Sequence: myseqs/S130-RVA-58.fastq</b>  len=352"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " HRVbo-f len=15 match=100% [172:186]\n",
      " HRVma-f len=19 match=89% [218:236]\n",
      "target          217 TTGACAAGGTGTGAAGAGC 236\n",
      "                  0 |.||||.||||||||||||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [256:273]\n",
      " HRVka5-ri (rev) len=18 match=100% [256:273]\n",
      " HRVbo-p len=18 match=100% [256:273]\n",
      " HRVkaV-fo len=20 match=100% [261:280]\n",
      " ENTng-f len=19 match=100% [262:280]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>S130-RVA-58 aligns to FJ445142.1[184:536]</b> match 94%"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ref = printSeqBinding(\"refseq/Rhinovirus-A58.fasta\", \"fasta\")\n",
    "seq = printSeqBinding(\"myseqs/S130-RVA-58.fastq\")\n",
    "printAlignment(ref, seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ed986c6",
   "metadata": {},
   "source": [
    "## Rhinovirus A-62: S142"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "06db7589",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b>Sequence: refseq/Rhinovirus-A62.fasta</b>  len=7131"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " ENTrc-f1 len=21 match=100% [163:183]\n",
      " HRVka5-f len=21 match=100% [163:183]\n",
      " HRVbo-f len=15 match=100% [353:367]\n",
      " HRVma-f len=19 match=89% [400:418]\n",
      "target          399 TTGACAAGGTGTGAAGAGC 418\n",
      "                  0 |.||||.||||||||||||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [438:455]\n",
      " HRVka5-ri (rev) len=18 match=100% [438:455]\n",
      " HRVbo-p len=18 match=100% [438:455]\n",
      " HRVkaV-fo len=20 match=100% [443:462]\n",
      " ENTng-f len=19 match=100% [444:462]\n",
      " HRVma-r (rev) len=20 match=95% [524:543]\n",
      "target          523 GGATGGGACCAACTACTTTG 543\n",
      "                  0 ||||||||||.|||||||||  20\n",
      "query             0 GGATGGGACCGACTACTTTG  20\n",
      "\n",
      " ENTng-p len=18 match=100% [526:543]\n",
      " ENTrc-r (rev) len=21 match=100% [535:555]\n",
      " HRVka5-ro (rev) len=23 match=100% [535:557]\n",
      " ENTng-r (rev) len=22 match=100% [536:557]\n",
      " HRVkaV-fi len=22 match=100% [536:557]\n",
      " HRVbo-r (rev) len=19 match=100% [539:557]\n",
      " HRVkaV-r (rev) len=23 match=100% [1050:1072]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>Sequence: myseqs/S142-RVA-62.fastq</b>  len=346"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " HRVbo-f len=15 match=100% [163:177]\n",
      " HRVma-f len=19 match=89% [209:227]\n",
      "target          208 TTGACAAGGTGTGAAGAGC 227\n",
      "                  0 |.||||.||||||||||||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [247:264]\n",
      " HRVka5-ri (rev) len=18 match=100% [247:264]\n",
      " HRVbo-p len=18 match=100% [247:264]\n",
      " HRVkaV-fo len=20 match=100% [252:271]\n",
      " ENTng-f len=19 match=100% [253:271]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>S142-RVA-62 aligns to FJ445145.1[191:537]</b> match 95%"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "target          191 GATACGCTCCAACAGGGCGAAAACAATYTAAATCGTTAACCGCAAAGTGACTACGCAAAG\n",
      "                  0 .||||||||||||||||||||||||.|..|.|||||||||||||||||||||||||||||\n",
      "query             0 TATACGCTCCAACAGGGCGAAAACAGTCAAGATCGTTAACCGCAAAGTGACTACGCAAAG\n",
      "\n",
      "target          251 CTTAGTAATACCTTGAAGAATTTATGGCTGGTCGTTCCGCTATAACCCC-TAGTAGACCT\n",
      "                 60 |||||||||.|||||||||.|.|||||||||||||||||||||||||||-||||||||||\n",
      "query            60 CTTAGTAATGCCTTGAAGAGTCTATGGCTGGTCGTTCCGCTATAACCCCCTAGTAGACCT\n",
      "\n",
      "target          310 GGCAGATGAGGCTAGAAATACCCCACTGGTAACAGTGTTCTAGCCTGCGTGGCTGCCTGC\n",
      "                120 ||||||||||||||||||.|||||||||||.|||||||||||||||||||||||||||||\n",
      "query           120 GGCAGATGAGGCTAGAAACACCCCACTGGTGACAGTGTTCTAGCCTGCGTGGCTGCCTGC\n",
      "\n",
      "target          370 ACACCCTTTTTGGGTGTGAAGCCATATATTTGACAAGGTGTGAAGAGCCCCGTGTGCTCA\n",
      "                180 |||||||||.-||||||||||||.||||||||||||||||||||||||||||||||||||\n",
      "query           180 ACACCCTTTC-GGGTGTGAAGCCGTATATTTGACAAGGTGTGAAGAGCCCCGTGTGCTCA\n",
      "\n",
      "target          430 CTTTGAGTCCTCCGGCCCCTGAATGTGGCTAACCTTAACCCTGCAGCTAGTGTATGTAAT\n",
      "                240 ||||||||||||||||||||||||||||||||||||||||||||||||||.|||||||||\n",
      "query           239 CTTTGAGTCCTCCGGCCCCTGAATGTGGCTAACCTTAACCCTGCAGCTAGCGTATGTAAT\n",
      "\n",
      "target          490 CCAACATATGGCTAGTCGTAATGAGCAATTGCGGGATGGGACCAACT 537\n",
      "                300 |||||||||.||||||||||||||||||||||||||||||||||||| 347\n",
      "query           299 CCAACATATTGCTAGTCGTAATGAGCAATTGCGGGATGGGACCAACT 346\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ref = printSeqBinding(\"refseq/Rhinovirus-A62.fasta\", \"fasta\")\n",
    "seq = printSeqBinding(\"myseqs/S142-RVA-62.fastq\")\n",
    "printAlignment(ref, seq, True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "749a46df",
   "metadata": {},
   "source": [
    "This is my first plasmidsaurus sequence. It's interesting that there's about 8 bases missing plus that of the whole primer sequence at the start, and just a couple bases overlap with the primer at the end. This is likely due to the use of rapid ONT chemistry which breaks the amplicons randomly with a transposon. There's also some very low quality bases at either end which I may need to trim for some applications. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20938e2c",
   "metadata": {},
   "source": [
    "## Rhinovirus A-68: S148"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3f30abef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b>Sequence: refseq/Rhinovirus-A68.fasta</b>  len=7164"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " ENTrc-f1 len=21 match=95% [165:185]\n",
      "target          164 CAAGCACTTCTGTTACCCCGG 185\n",
      "                  0 ||||||||||||||.||||||  21\n",
      "query             0 CAAGCACTTCTGTTTCCCCGG  21\n",
      "\n",
      " HRVka5-f len=21 match=95% [165:185]\n",
      "target          164 CAAGCACTTCTGTTACCCCGG 185\n",
      "                  0 ||||||||||||||.||||||  21\n",
      "query             0 CAAGCACTTCTGTTTCCCCGG  21\n",
      "\n",
      " HRVbo-f len=15 match=100% [357:371]\n",
      " HRVma-f len=19 match=89% [405:423]\n",
      "target          404 TTGACAAGGTGTGAAGAGC 423\n",
      "                  0 |.||||.||||||||||||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [443:460]\n",
      " HRVka5-ri (rev) len=18 match=100% [443:460]\n",
      " HRVbo-p len=18 match=100% [443:460]\n",
      " HRVkaV-fo len=20 match=100% [448:467]\n",
      " ENTng-f len=19 match=100% [449:467]\n",
      " HRVma-r (rev) len=20 match=95% [529:548]\n",
      "target          528 GGATGGGACCAACTACTTTG 548\n",
      "                  0 ||||||||||.|||||||||  20\n",
      "query             0 GGATGGGACCGACTACTTTG  20\n",
      "\n",
      " ENTng-p len=18 match=100% [531:548]\n",
      " ENTrc-r (rev) len=21 match=100% [540:560]\n",
      " HRVka5-ro (rev) len=23 match=100% [540:562]\n",
      " ENTng-r (rev) len=22 match=100% [541:562]\n",
      " HRVkaV-fi len=22 match=100% [541:562]\n",
      " HRVbo-r (rev) len=19 match=100% [544:562]\n",
      " HRVkaV-r (rev) len=23 match=100% [1058:1080]\n",
      " random len=15 match=80% [1364:1377]\n",
      "  Primer mismatch in the last 1 bases\n",
      "target         1363 CTACTAGGGAACATA 1378\n",
      "                  0 |||||.|||.||||.   15\n",
      "query             0 CTACTGGGGGACATT   15\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>Sequence: myseqs/S148-RVA-68.fastq</b>  len=349"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " HRVbo-f len=15 match=100% [178:192]\n",
      " HRVma-f len=19 match=89% [226:244]\n",
      "target          225 TTGACAAGGTGTGAAGAGC 244\n",
      "                  0 |.||||.||||||||||||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [264:281]\n",
      " HRVka5-ri (rev) len=18 match=100% [264:281]\n",
      " HRVbo-p len=18 match=100% [264:281]\n",
      " HRVkaV-fo len=20 match=100% [269:288]\n",
      " ENTng-f len=19 match=100% [270:288]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>S148-RVA-68 aligns to FJ445150.1[179:529]</b> match 93%"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ref = printSeqBinding(\"refseq/Rhinovirus-A68.fasta\", \"fasta\")\n",
    "seq = printSeqBinding(\"myseqs/S148-RVA-68.fastq\")\n",
    "printAlignment(ref, seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6cf25b65",
   "metadata": {},
   "source": [
    "## Rhinovirus A-77: S59"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "30a1c993",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b>Sequence: refseq/Rhinovirus-A77.gb</b>  len=7136"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " ENTrc-f1 len=21 match=100% [163:183]\n",
      " HRVka5-f len=21 match=100% [163:183]\n",
      " HRVbo-f len=15 match=100% [354:368]\n",
      " HRVma-f len=19 match=89% [400:418]\n",
      "target          399 TTGACAAGGTGTGAAGAGC 418\n",
      "                  0 |.||||.||||||||||||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [438:455]\n",
      " HRVka5-ri (rev) len=18 match=100% [438:455]\n",
      " HRVbo-p len=18 match=100% [438:455]\n",
      " HRVkaV-fo len=20 match=100% [443:462]\n",
      " ENTng-f len=19 match=100% [444:462]\n",
      " HRVma-r (rev) len=20 match=95% [524:543]\n",
      "target          523 GGATGGGACCAACTACTTTG 543\n",
      "                  0 ||||||||||.|||||||||  20\n",
      "query             0 GGATGGGACCGACTACTTTG  20\n",
      "\n",
      " ENTng-p len=18 match=100% [526:543]\n",
      " ENTrc-r (rev) len=21 match=100% [535:555]\n",
      " HRVka5-ro (rev) len=23 match=100% [535:557]\n",
      " ENTng-r (rev) len=22 match=100% [536:557]\n",
      " HRVkaV-fi len=22 match=100% [536:557]\n",
      " HRVbo-r (rev) len=19 match=100% [539:557]\n",
      " HRVkaV-r (rev) len=23 match=100% [1049:1071]\n",
      " HRVkaV-r (rev) len=23 match=83% [4273:4295]\n",
      "target         1048 GGATGGTGGTGGAAATTGCCAGA 1071\n",
      "                  0 ||.|||||.|||||..|.||.||   23\n",
      "query             0 GGNTGGTGBTGGAARYTNCCNGA   23\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>Sequence: myseqs/S59-RVA-77.fastq</b>  len=351"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " HRVbo-f len=15 match=100% [171:185]\n",
      " HRVma-f len=19 match=89% [217:235]\n",
      "target          216 TTGACAAGGTGTGAAGAGC 235\n",
      "                  0 |.||||.||||||||||||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [255:272]\n",
      " HRVka5-ri (rev) len=18 match=100% [255:272]\n",
      " HRVbo-p len=18 match=100% [255:272]\n",
      " HRVkaV-fo len=20 match=100% [260:279]\n",
      " ENTng-f len=19 match=100% [261:279]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>S59-RVA-77 aligns to FJ445154[183:534]</b> match 96%"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ref = printSeqBinding(\"refseq/Rhinovirus-A77.gb\", \"gb\")\n",
    "seq = printSeqBinding(\"myseqs/S59-RVA-77.fastq\")\n",
    "printAlignment(ref, seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c95ec11c",
   "metadata": {},
   "source": [
    "## Rhinovirus B-27: S153"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8150622f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b>Sequence: refseq/Rhinovirus-B27.fasta</b>  len=7217"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " ENTrc-f1 len=21 match=100% [180:200]\n",
      " HRVka5-f len=21 match=100% [180:200]\n",
      " HRVbo-f len=15 match=100% [368:382]\n",
      " HRVma-p len=18 match=100% [451:468]\n",
      " HRVka5-ri (rev) len=18 match=100% [451:468]\n",
      " HRVbo-p len=18 match=100% [451:468]\n",
      " HRVkaV-fo len=20 match=100% [456:475]\n",
      " ENTng-f len=19 match=100% [457:475]\n",
      " HRVma-r (rev) len=20 match=100% [537:556]\n",
      " ENTng-p len=18 match=100% [539:556]\n",
      " ENTrc-r (rev) len=21 match=100% [548:568]\n",
      " HRVka5-ro (rev) len=23 match=100% [548:570]\n",
      " ENTng-r (rev) len=22 match=100% [549:570]\n",
      " HRVkaV-fi len=22 match=100% [549:570]\n",
      " HRVbo-r (rev) len=19 match=100% [552:570]\n",
      " HRVkaV-r (rev) len=23 match=100% [1063:1085]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>Sequence: myseqs/S153-RVB-27.fastq</b>  len=333"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " HRVbo-f len=15 match=100% [173:187]\n",
      " HRVma-f len=19 match=89% [217:235]\n",
      "target          216 TGGACATGGTGTGAAGACC 235\n",
      "                  0 ||||||.||||||||||.|  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [256:273]\n",
      " HRVka5-ri (rev) len=18 match=100% [256:273]\n",
      " HRVbo-p len=18 match=100% [256:273]\n",
      " HRVkaV-fo len=20 match=100% [261:280]\n",
      " ENTng-f len=19 match=100% [262:280]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>S153-RVB-27 aligns to FJ445186.1[195:528]</b> match 95%"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ref = printSeqBinding(\"refseq/Rhinovirus-B27.fasta\", \"fasta\")\n",
    "seq = printSeqBinding(\"myseqs/S153-RVB-27.fastq\")\n",
    "printAlignment(ref, seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8128028a",
   "metadata": {},
   "source": [
    "## Rhinovirus C-1: S48, S94"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "69dfacc6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b>Sequence: refseq/Rhinovirus-C1.gb</b>  len=6944"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " ENTrc-f1 len=21 match=100% [28:48]\n",
      " HRVka5-f len=21 match=100% [28:48]\n",
      " HRVbo-f len=15 match=100% [220:234]\n",
      " HRVma-f len=19 match=95% [266:284]\n",
      "target          265 TGGACAAGGTGTGAAGAGC 284\n",
      "                  0 ||||||.||||||||||||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [304:321]\n",
      " HRVka5-ri (rev) len=18 match=100% [304:321]\n",
      " HRVbo-p len=18 match=100% [304:321]\n",
      " HRVkaV-fo len=20 match=100% [309:328]\n",
      " ENTng-f len=19 match=100% [310:328]\n",
      " HRVma-r (rev) len=20 match=90% [390:409]\n",
      "target          389 GGATGGAACCAACTACTTTG 409\n",
      "                  0 ||||||.|||.|||||||||  20\n",
      "query             0 GGATGGGACCGACTACTTTG  20\n",
      "\n",
      " ENTng-p len=18 match=100% [392:409]\n",
      " ENTrc-r (rev) len=21 match=100% [401:421]\n",
      " HRVka5-ro (rev) len=23 match=100% [401:423]\n",
      " ENTng-r (rev) len=22 match=100% [402:423]\n",
      " HRVkaV-fi len=22 match=100% [402:423]\n",
      " HRVbo-r (rev) len=19 match=100% [405:423]\n",
      " HRVkaV-r (rev) len=23 match=100% [908:930]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>Sequence: myseqs/S48-RVC-1.fastq</b>  len=352"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " HRVbo-f len=15 match=100% [172:186]\n",
      " HRVma-f len=19 match=89% [218:235]\n",
      "  Primer mismatch in the last 1 bases\n",
      "target          217 TGGACAAGGTGTGAAGAGT 236\n",
      "                  0 ||||||.|||||||||||.  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [256:273]\n",
      " HRVka5-ri (rev) len=18 match=100% [256:273]\n",
      " HRVbo-p len=18 match=100% [256:273]\n",
      " HRVkaV-fo len=20 match=100% [261:280]\n",
      " ENTng-f len=19 match=100% [262:280]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>Sequence: myseqs/S94-RVC-1.fastq</b>  len=352"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " HRVbo-f len=15 match=100% [172:186]\n",
      " HRVma-f len=19 match=89% [218:235]\n",
      "  Primer mismatch in the last 1 bases\n",
      "target          217 TGGACAAGGTGTGAAGAGT 236\n",
      "                  0 ||||||.|||||||||||.  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [256:273]\n",
      " HRVka5-ri (rev) len=18 match=100% [256:273]\n",
      " HRVbo-p len=18 match=100% [256:273]\n",
      " HRVkaV-fo len=20 match=100% [261:280]\n",
      " ENTng-f len=19 match=100% [262:280]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>S48-RVC-1 aligns to EF077279[48:400]</b> match 98%"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<b>S94-RVC-1 aligns to EF077279[48:400]</b> match 98%"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<b>S94-RVC-1 aligns to S48-RVC-1[0:352]</b> match 99%"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "target            0 TACCCTCGTATACGCTTCACCCGAGGCGAAAAATGAGGTTATCGTTACCCGCAAAGTGCC\n",
      "                  0 ||||||.|||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "query             0 TACCCTTGTATACGCTTCACCCGAGGCGAAAAATGAGGTTATCGTTACCCGCAAAGTGCC\n",
      "\n",
      "target           60 TACGAGAAACCTAGTAGCATTTTTGAAGCCTATGGTTGGTCGCTCAACTGTTTACCCAGC\n",
      "                 60 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "query            60 TACGAGAAACCTAGTAGCATTTTTGAAGCCTATGGTTGGTCGCTCAACTGTTTACCCAGC\n",
      "\n",
      "target          120 AGTAGACCTGGCAGATGAGGCTAGATGTTCCCCACCAGCGATGGTGATCTAGCCTGCGTG\n",
      "                120 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "query           120 AGTAGACCTGGCAGATGAGGCTAGATGTTCCCCACCAGCGATGGTGATCTAGCCTGCGTG\n",
      "\n",
      "target          180 GCTGCCTGCACACTCTATTGAGTGTGAAGCCAGAAAGTGGACAAGGTGTGAAGAGTCTAT\n",
      "                180 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "query           180 GCTGCCTGCACACTCTATTGAGTGTGAAGCCAGAAAGTGGACAAGGTGTGAAGAGTCTAT\n",
      "\n",
      "target          240 TGTGCTCACTTTGAGTCCTCCGGCCCCTGAATGTGGCTAATCCTAACCCCGTAGCTGTTG\n",
      "                240 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "query           240 TGTGCTCACTTTGAGTCCTCCGGCCCCTGAATGTGGCTAATCCTAACCCCGTAGCTGTTG\n",
      "\n",
      "target          300 CATGTAATCCAACATGTCTGCAGTCGTAATGGGCAACTATGGGATGGAACCA 352\n",
      "                300 |||||||||||||||||||||||||||||||||||||||||||||||||||| 352\n",
      "query           300 CATGTAATCCAACATGTCTGCAGTCGTAATGGGCAACTATGGGATGGAACCA 352\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ref = printSeqBinding(\"refseq/Rhinovirus-C1.gb\", \"gb\")\n",
    "seq1 = printSeqBinding(\"myseqs/S48-RVC-1.fastq\")\n",
    "seq2 = printSeqBinding(\"myseqs/S94-RVC-1.fastq\")\n",
    "\n",
    "printAlignment(ref, seq1)\n",
    "printAlignment(ref, seq2)\n",
    "printAlignment(seq1, seq2, printAlignment=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b2014b1",
   "metadata": {},
   "source": [
    "## Rhinovirus C-44: S147"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "24bf89ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b>Sequence: refseq/Rhinovirus-C44.fasta</b>  len=7098"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " ENTrc-f1 len=21 match=95% [135:155]\n",
      "target          134 CAAGCACTTCTGTTACCCCGG 155\n",
      "                  0 ||||||||||||||.||||||  21\n",
      "query             0 CAAGCACTTCTGTTTCCCCGG  21\n",
      "\n",
      " HRVka5-f len=21 match=95% [135:155]\n",
      "target          134 CAAGCACTTCTGTTACCCCGG 155\n",
      "                  0 ||||||||||||||.||||||  21\n",
      "query             0 CAAGCACTTCTGTTTCCCCGG  21\n",
      "\n",
      " HRVbo-f len=15 match=100% [331:345]\n",
      " HRVma-f len=19 match=89% [373:391]\n",
      "target          372 TTGACAAGGTGTGAAGAGC 391\n",
      "                  0 |.||||.||||||||||||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [411:428]\n",
      " HRVka5-ri (rev) len=18 match=100% [411:428]\n",
      " HRVbo-p len=18 match=100% [411:428]\n",
      " HRVkaV-fo len=20 match=100% [416:435]\n",
      " ENTng-f len=19 match=100% [417:435]\n",
      " HRVma-r (rev) len=20 match=90% [496:515]\n",
      "target          495 GGACGGAACCGACTACTTTG 515\n",
      "                  0 |||.||.|||||||||||||  20\n",
      "query             0 GGATGGGACCGACTACTTTG  20\n",
      "\n",
      " ENTng-p len=18 match=100% [498:515]\n",
      " ENTrc-r (rev) len=21 match=100% [507:527]\n",
      " HRVka5-ro (rev) len=23 match=100% [507:529]\n",
      " ENTng-r (rev) len=22 match=100% [508:529]\n",
      " HRVkaV-fi len=22 match=100% [508:529]\n",
      " HRVbo-r (rev) len=19 match=100% [511:529]\n",
      " HRVkaV-r (rev) len=23 match=100% [1017:1039]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>Sequence: myseqs/S147-RVC-44.fastq</b>  len=338"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " HRVbo-f len=15 match=100% [158:172]\n",
      " HRVma-f len=19 match=89% [200:218]\n",
      "target          199 TTGACAAGGTGTGAAGAGC 218\n",
      "                  0 |.||||.||||||||||||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [238:255]\n",
      " HRVka5-ri (rev) len=18 match=100% [238:255]\n",
      " HRVbo-p len=18 match=100% [238:255]\n",
      " HRVkaV-fo len=20 match=100% [243:262]\n",
      " ENTng-f len=19 match=100% [244:262]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>S147-RVC-44 aligns to MZ153259.1[173:512]</b> match 97%"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "target          173 TTCCCCAAGACCGAAGCCTTGACTGTCGTTACCCGCACAACTACTGGACAAAGCCTAGTA\n",
      "                  0 .|||||||||||||||||||||.||||||||||||||||.||||||||||||||||||||\n",
      "query             0 GTCCCCAAGACCGAAGCCTTGATTGTCGTTACCCGCACAGCTACTGGACAAAGCCTAGTA\n",
      "\n",
      "target          233 ATACCTGACGATATAGTGTGGTTGGTCGCTCCACCAGTGAACCCCTGGTAGACCTGGCAG\n",
      "                 60 ||||||||||||||||.|||||||||||||||||||||||||||||||||||||||||||\n",
      "query            60 ATACCTGACGATATAGCGTGGTTGGTCGCTCCACCAGTGAACCCCTGGTAGACCTGGCAG\n",
      "\n",
      "target          293 ATGAGGCTGGAGTTTCCCCACTGGTAACAGTGTTCCAGCCTGCGTGGCTGCCTGCTCCCT\n",
      "                120 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "query           120 ATGAGGCTGGAGTTTCCCCACTGGTAACAGTGTTCCAGCCTGCGTGGCTGCCTGCTCCCT\n",
      "\n",
      "target          353 CACGGGAGAAGCCATTCTATTGACAAGGTGTGAAGAGCCCCGTACGCTAGTTGTGAGTCC\n",
      "                180 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "query           180 CACGGGAGAAGCCATTCTATTGACAAGGTGTGAAGAGCCCCGTACGCTAGTTGTGAGTCC\n",
      "\n",
      "target          413 TCCGGCCCCTGAATGCGGCTAATCCAACCCTGCAGCCATTGCATGGAAGCCACCATGTGG\n",
      "                240 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||.|\n",
      "query           240 TCCGGCCCCTGAATGCGGCTAATCCAACCCTGCAGCCATTGCATGGAAGCCACCATGTAG\n",
      "\n",
      "target          473 ATGGTCGTAATGAGTAATTGCAGGACGGAACCGACTACT 512\n",
      "                300 |||||||||||||||||||||.||||||||||||||-|| 339\n",
      "query           300 ATGGTCGTAATGAGTAATTGCGGGACGGAACCGACT-CT 338\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ref = printSeqBinding(\"refseq/Rhinovirus-C44.fasta\", \"fasta\")\n",
    "seq = printSeqBinding(\"myseqs/S147-RVC-44.fastq\")\n",
    "printAlignment(ref, seq, True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aca103e9",
   "metadata": {},
   "source": [
    "## Rhinovirus C-45: S69, S91"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6a56d87c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b>Sequence: refseq/Rhinovirus-C45.fasta</b>  len=6972"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " ENTrc-f1 len=21 match=90% [26:46]\n",
      "target           25 CAAACACTTCTGTTCCCCCGG 46\n",
      "                  0 |||.||||||||||.|||||| 21\n",
      "query             0 CAAGCACTTCTGTTTCCCCGG 21\n",
      "\n",
      " HRVka5-f len=21 match=90% [26:46]\n",
      "target           25 CAAACACTTCTGTTCCCCCGG 46\n",
      "                  0 |||.||||||||||.|||||| 21\n",
      "query             0 CAAGCACTTCTGTTTCCCCGG 21\n",
      "\n",
      " HRVbo-f len=15 match=100% [220:234]\n",
      " HRVma-f len=19 match=95% [267:285]\n",
      "target          266 TAGACAGGGTGTGAAGAGC 285\n",
      "                  0 |.|||||||||||||||||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [305:322]\n",
      " HRVka5-ri (rev) len=18 match=100% [305:322]\n",
      " HRVbo-p len=18 match=100% [305:322]\n",
      " HRVkaV-fo len=20 match=100% [310:329]\n",
      " ENTng-f len=19 match=100% [311:329]\n",
      " HRVma-r (rev) len=20 match=95% [390:409]\n",
      "target          389 GGATGGAACCGACTACTTTG 409\n",
      "                  0 ||||||.|||||||||||||  20\n",
      "query             0 GGATGGGACCGACTACTTTG  20\n",
      "\n",
      " ENTng-p len=18 match=100% [392:409]\n",
      " ENTrc-r (rev) len=21 match=100% [401:421]\n",
      " HRVka5-ro (rev) len=23 match=100% [401:423]\n",
      " ENTng-r (rev) len=22 match=100% [402:423]\n",
      " HRVkaV-fi len=22 match=100% [402:423]\n",
      " HRVbo-r (rev) len=19 match=100% [405:423]\n",
      " HRVkaV-r (rev) len=23 match=96% [910:932]\n",
      "target          909 GGATGGTGGTGGAAAATACCTGA 932\n",
      "                  0 ||.|||||.|||||..|.||.||  23\n",
      "query             0 GGNTGGTGBTGGAARYTNCCNGA  23\n",
      "\n",
      " HRVbo-f (rev) len=15 match=80% [2074:2089]\n",
      "target         2073 GGCAGTCAGGGCGGGC 2089\n",
      "                  0 |||.|.||.|-|.|||   16\n",
      "query             0 GGCMGCCACG-CRGGC   15\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>Sequence: myseqs/S69-RVC-45.fastq</b>  len=354"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " HRVbo-f len=15 match=100% [174:188]\n",
      " HRVma-f len=19 match=95% [221:239]\n",
      "target          220 TAGACAGGGTGTGAAGAGC 239\n",
      "                  0 |.|||||||||||||||||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [259:276]\n",
      " HRVka5-ri (rev) len=18 match=100% [259:276]\n",
      " HRVbo-p len=18 match=100% [259:276]\n",
      " HRVkaV-fo len=20 match=100% [264:283]\n",
      " ENTng-f len=19 match=100% [265:283]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>Sequence: myseqs/S91-RVC-45.fastq</b>  len=354"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " HRVbo-f len=15 match=100% [174:188]\n",
      " HRVma-f len=19 match=95% [221:239]\n",
      "target          220 TAGACAGGGTGTGAAGAGC 239\n",
      "                  0 |.|||||||||||||||||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [259:276]\n",
      " HRVka5-ri (rev) len=18 match=100% [259:276]\n",
      " HRVbo-p len=18 match=100% [259:276]\n",
      " HRVkaV-fo len=20 match=100% [264:283]\n",
      " ENTng-f len=19 match=100% [265:283]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>S69-RVC-45 aligns to KY624849.1[46:400]</b> match 98%"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<b>S91-RVC-45 aligns to KY624849.1[46:400]</b> match 98%"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<b>S91-RVC-45 aligns to S69-RVC-45[0:354]</b> match 99%"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "target            0 CATGTGAGGAATAGGCTCCAAAAGGGCTAAAGCCACTAGTGTCGTTATCCGCATTGGTAC\n",
      "                  0 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "query             0 CATGTGAGGAATAGGCTCCAAAAGGGCTAAAGCCACTAGTGTCGTTATCCGCATTGGTAC\n",
      "\n",
      "target           60 TACGCAAAGCCTAGTATCACTCTGGAAGTCTCTCGGTTGGTCGCTCCACCAGCTACCCCA\n",
      "                 60 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "query            60 TACGCAAAGCCTAGTATCACTCTGGAAGTCTCTCGGTTGGTCGCTCCACCAGCTACCCCA\n",
      "\n",
      "target          120 CTGGTAGACCTGGCAGATGAGGCAGGACTCTCCCCACTGGTGACAGTGGTCCTGCCTGCG\n",
      "                120 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "query           120 CTGGTAGACCTGGCAGATGAGGCAGGACTCTCCCCACTGGTGACAGTGGTCCTGCCTGCG\n",
      "\n",
      "target          180 TGGCTGCCTGCACACCCCTACGGGGTGTGAAGCCTAGAGATAGACAGGGTGTGAAGAGCC\n",
      "                180 |||||||||||||||||||||.||||||||||||||||||||||||||||||||||||||\n",
      "query           180 TGGCTGCCTGCACACCCCTACAGGGTGTGAAGCCTAGAGATAGACAGGGTGTGAAGAGCC\n",
      "\n",
      "target          240 CCGTGTGCTACCAATGAGTCCTCCGGCCCCTGAATGCGGCTAATCTTACCCCACAGCTGT\n",
      "                240 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "query           240 CCGTGTGCTACCAATGAGTCCTCCGGCCCCTGAATGCGGCTAATCTTACCCCACAGCTGT\n",
      "\n",
      "target          300 TGCACGCAGTCCAGCGTGTATGCAGTCGTAATGAGCAATTGTGGGACGGAACCG 354\n",
      "                300 |||||||||||||||||||||||||||||||||||||||||||||||||||||| 354\n",
      "query           300 TGCACGCAGTCCAGCGTGTATGCAGTCGTAATGAGCAATTGTGGGACGGAACCG 354\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ref = printSeqBinding(\"refseq/Rhinovirus-C45.fasta\", \"fasta\")\n",
    "seq1 = printSeqBinding(\"myseqs/S69-RVC-45.fastq\")\n",
    "seq2 = printSeqBinding(\"myseqs/S91-RVC-45.fastq\")\n",
    "\n",
    "printAlignment(ref, seq1)\n",
    "printAlignment(ref, seq2)\n",
    "printAlignment(seq1, seq2, printAlignment=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af89317f",
   "metadata": {},
   "source": [
    "# Other reference sequences\n",
    "## RefSeq Rhinovirus A-89\n",
    "\n",
    "Test against the full A-89 genome since this is what's often used as the common coordinate system"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6bac9a3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b>Sequence: refseq/Rhinovirus-A89.gb</b>  len=7152"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " ENTrc-f1 len=21 match=100% [165:185]\n",
      " HRVka5-f len=21 match=100% [165:185]\n",
      " HRVbo-f len=15 match=100% [357:371]\n",
      " HRVma-f len=19 match=89% [404:422]\n",
      "target          403 TTGACAAGGTGTGAAGAGC 422\n",
      "                  0 |.||||.||||||||||||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [442:459]\n",
      " HRVka5-ri (rev) len=18 match=100% [442:459]\n",
      " HRVbo-p len=18 match=100% [442:459]\n",
      " HRVkaV-fo len=20 match=100% [447:466]\n",
      " ENTng-f len=19 match=100% [448:466]\n",
      " HRVma-r (rev) len=20 match=95% [528:547]\n",
      "target          527 GGATGGGACCAACTACTTTG 547\n",
      "                  0 ||||||||||.|||||||||  20\n",
      "query             0 GGATGGGACCGACTACTTTG  20\n",
      "\n",
      " ENTng-p len=18 match=100% [530:547]\n",
      " ENTrc-r (rev) len=21 match=100% [539:559]\n",
      " HRVka5-ro (rev) len=23 match=100% [539:561]\n",
      " ENTng-r (rev) len=22 match=100% [540:561]\n",
      " HRVkaV-fi len=22 match=100% [540:561]\n",
      " HRVbo-r (rev) len=19 match=100% [543:561]\n",
      " HRVkaV-r (rev) len=23 match=100% [1054:1076]\n",
      " random (rev) len=15 match=80% [2786:2800]\n",
      "  Primer mismatch in first 1 bases\n",
      "target         2784 TATGTACCCCCAGGAG 2800\n",
      "                  0 .||||-|||||||.||   16\n",
      "query             0 AATGT-CCCCCAGTAG   15\n",
      "\n"
     ]
    }
   ],
   "source": [
    "printSeqBinding(\"refseq/Rhinovirus-A89.gb\", \"gb\");"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "754322cf",
   "metadata": {},
   "source": [
    "Overall this isn't anything too special, no compelling reason to switch from HRV Ka."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2ebacb3",
   "metadata": {},
   "source": [
    "## RefSeq Rhinovirus C-11\n",
    "\n",
    "Test against the full C-11 genome since I have several hits to it. There's a few mismatches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7aa394f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b>Sequence: refseq/Rhinovirus-C11.fasta</b>  len=7108"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " ENTrc-f1 len=21 match=90% [163:183]\n",
      "target          162 CAAATACTTCTGTTTCCCCGG 183\n",
      "                  0 |||..||||||||||||||||  21\n",
      "query             0 CAAGCACTTCTGTTTCCCCGG  21\n",
      "\n",
      " HRVka5-f len=21 match=90% [163:183]\n",
      "target          162 CAAATACTTCTGTTTCCCCGG 183\n",
      "                  0 |||..||||||||||||||||  21\n",
      "query             0 CAAGCACTTCTGTTTCCCCGG  21\n",
      "\n",
      " HRVbo-f len=15 match=100% [357:371]\n",
      " HRVma-f len=19 match=84% [407:423]\n",
      "  Primer mismatch in first 2 bases\n",
      "target          404 CAGACAAGGTGTGAAGAGC 423\n",
      "                  0 ..||||.||||||||||||  19\n",
      "query             0 TGGACAGGGTGTGAAGAGC  19\n",
      "\n",
      " HRVma-p len=18 match=100% [443:460]\n",
      " HRVka5-ri (rev) len=18 match=100% [443:460]\n",
      " HRVbo-p len=18 match=100% [443:460]\n",
      " HRVkaV-fo len=20 match=100% [448:467]\n",
      " ENTng-f len=19 match=100% [449:467]\n",
      " HRVma-r (rev) len=20 match=95% [528:547]\n",
      "target          527 GGATGGAACCGACTACTTTG 547\n",
      "                  0 ||||||.|||||||||||||  20\n",
      "query             0 GGATGGGACCGACTACTTTG  20\n",
      "\n",
      " ENTng-p len=18 match=100% [530:547]\n",
      " ENTrc-r (rev) len=21 match=100% [539:559]\n",
      " HRVka5-ro (rev) len=23 match=100% [539:561]\n",
      " ENTng-r (rev) len=22 match=100% [540:561]\n",
      " HRVkaV-fi len=22 match=100% [540:561]\n",
      " HRVbo-r (rev) len=19 match=100% [543:561]\n",
      " HRVkaV-r (rev) len=23 match=96% [1048:1070]\n",
      "target         1047 GGCTGGTGGTGGAAAATTCCAGA 1070\n",
      "                  0 ||.|||||.|||||..|.||.||   23\n",
      "query             0 GGNTGGTGBTGGAARYTNCCNGA   23\n",
      "\n"
     ]
    }
   ],
   "source": [
    "printSeqBinding(\"refseq/Rhinovirus-C11.fasta\", \"fasta\");"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b358a6c7",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "fb916a3b",
   "metadata": {},
   "source": [
    "# Wisdom primers\n",
    "\n",
    "See how the primers from this older paper compare: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2786677/\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b3f54b8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading primers: HRV-Wisdom-primers.fasta\n",
      "  HRV-Wi-5UTR-fo (48 variations)\n",
      "  HRV-Wi-5UTR-ro\n",
      "  HRV-Wi-5UTR-f1i (16 variations)\n",
      "  HRV-Wi-5UTR-f2i (4 variations)\n",
      "  HRV-Wi-5UTR-f3i (4 variations)\n",
      "  HRV-Wi-5UTR-ri (2 variations)\n",
      "  HRV-Wi-VP4VP2-fo (2 variations)\n",
      "  HRV-Wi-VP4VP2-fi (2 variations)\n",
      "  HRV-Wi-VP4VP2-ri (192 variations)\n",
      "  HRV-Wi-VP4VP2-ro (768 variations)\n",
      "  HRV-Wi-VP1-fo (96 variations)\n",
      "  HRV-Wi-VP1-fi (64 variations)\n",
      "  HRV-Wi-VP1-ri (96 variations)\n",
      "  HRV-Wi-VP1-ro (48 variations)\n",
      "Read 1343 primers\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>Sequence: refseq/Rhinovirus-A89.gb</b>  len=7152"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " HRV-Wi-5UTR-fo.39 len=22 match=100% [164:185]\n",
      " HRV-Wi-5UTR-f1i.7 len=21 match=100% [354:374]\n",
      " HRV-Wi-5UTR-ri.0 (rev) len=21 match=100% [446:466]\n",
      " HRV-Wi-5UTR-ro (rev) len=23 match=100% [539:561]\n",
      " HRV-Wi-VP4VP2-ri.46 (rev) len=23 match=100% [1054:1076]\n",
      " HRV-Wi-VP4VP2-ro.231 (rev) len=25 match=100% [1090:1114]\n",
      " HRV-Wi-VP1-fo.75 len=23 match=100% [1966:1988]\n",
      " HRV-Wi-VP1-fi.56 len=19 match=89% [2441:2459]\n",
      "target         2440 TGGATGCTGCGGAAACCGG 2459\n",
      "                  0 ||||||||||.|||||.||   19\n",
      "query             0 TGGATGCTGCAGAAACAGG   19\n",
      "\n",
      " HRV-Wi-VP1-ri.94 (rev) len=26 match=96% [3319:3344]\n",
      "target         3318 GATCTAATCATATATCGAACAAACAC 3344\n",
      "                  0 ||||||||||||||.|||||||||||   26\n",
      "query             0 GATCTAATCATATACCGAACAAACAC   26\n",
      "\n",
      " HRV-Wi-VP1-ro.21 (rev) len=23 match=100% [3514:3536]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HRV-Wi-5UTR</th>\n",
       "      <th>HRV-Wi-VP4VP2</th>\n",
       "      <th>HRV-Wi-VP1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A89</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-C1</th>\n",
       "      <td>100</td>\n",
       "      <td>96</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A56</th>\n",
       "      <td>100</td>\n",
       "      <td>77</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A23</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhinovirus-A77</th>\n",
       "      <td>95</td>\n",
       "      <td>100</td>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                HRV-Wi-5UTR  HRV-Wi-VP4VP2  HRV-Wi-VP1\n",
       "Rhinovirus-A89          100            100          85\n",
       "Rhinovirus-C1           100             96          83\n",
       "Rhinovirus-A56          100             77         100\n",
       "Rhinovirus-A23          100            100          89\n",
       "Rhinovirus-A77           95            100          96"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wisdomPrimers = RCUtils.readPrimers(\"HRV-Wisdom-primers.fasta\", display=True)\n",
    "printSeqBinding(\"refseq/Rhinovirus-A89.gb\", \"gb\", primers=wisdomPrimers)\n",
    "\n",
    "summaryTable(\"refseq/Rhinovirus-*.gb\", \"gb\", primers=wisdomPrimers)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7059639",
   "metadata": {},
   "source": [
    "# Bochkov primers\n",
    "\n",
    "See how the primers from this 2014 paper compare: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2786677/\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "4c6375e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading primers: HRV-Bochkov-primers.fasta\n",
      "  HRV-Bv-5UTRn-A1 (degeneracy 2)\n",
      "  HRV-Bv-5UTRn-A2\n",
      "  HRV-Bv-5UTRn-A3\n",
      "  HRV-Bv-5UTRn-B1\n",
      "  HRV-Bv-5UTRn-B2\n",
      "  HRV-Bv-5UTRn-Cc (degeneracy 4)\n",
      "  HRV-Bv-5UTR-rev\n",
      "  HRV-Bv-5UTR-revseq\n",
      "  HRV-Bv-A53-f\n",
      "  HRV-Bv-B6-f\n",
      "  HRV-Bv-A53-r\n",
      "  HRV-Bv-B6-R1164-r\n",
      "  HRV-Bv-A82-f\n",
      "  HRV-Bv-B27-f\n",
      "  HRV-Bv-A82-r\n",
      "  HRV-Bv-B27-r\n",
      "  HRV-Bv-A21-f\n",
      "  HRV-Bv-A21-r\n",
      "  HRV-Bv-B6-R1912-r\n",
      "  HRV-Bv-5UTR-f (degeneracy 2)\n",
      "  HRV-Bv-VP2-C252-5r (degeneracy 2)\n",
      "Read 21 primers\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<b>Sequence: refseq/Rhinovirus-C1.fasta</b>  len=6944"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " HRV-Bv-5UTRn-A1 len=19 match=100% [31:49]\n",
      " HRV-Bv-A21-f len=20 match=95% [89:107]\n",
      "  Primer mismatch in first 1 bases\n",
      "target           87 TATCGTTATCCGCAAAGTGC 107\n",
      "                  0 .|||||||||||||||||||  20\n",
      "query             0 GATCGTTATCCGCAAAGTGC  20\n",
      "\n",
      " HRV-Bv-5UTR-revseq (rev) len=15 match=100% [304:318]\n",
      " HRV-Bv-5UTR-rev (rev) len=18 match=100% [401:418]\n",
      " HRV-Bv-VP2-C252-5r (rev) len=21 match=100% [710:730]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('GTGACAATCAGCCAGATTGTTAACGGTCAAGCACTTCTGTTTCCCCGGTACCCT...TTT'), id='EF077279.1', name='EF077279.1', description='EF077279.1 Rhinovirus C1 isolate NAT001 polyprotein gene, complete cds', dbxrefs=[])"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bochkovPrimers = RCUtils.readPrimers(\"HRV-Bochkov-primers.fasta\", display=True, expandAmbiguous=False)\n",
    "printSeqBinding(\"refseq/Rhinovirus-C1.fasta\", \"fasta\", primers=bochkovPrimers)\n",
    "\n",
    "#summaryTable(\"refseq/Rhinovirus-*.fasta\", \"fasta\", primers=bochkovPrimers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36511fb4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "0ba0ceb6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m = iupac_subst();\n",
    "m[\"I\"][\"C\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2124eaf3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "newpython",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
